plugin/kubernetes: filter ExternalName service queries for subdomains of subdomains

emaildanwilson opened the pull request:
### 1. Why is this pull request needed and what does it do?
This prevents queries for `*.*.service.namespace.svc.cluster.local` from being resolved when the service is type ExternalName. The `*` is not for a wildcard DNS query here, just to indicate that any valid dns label can be used in the query.

For example, a query for `www.google.com.default.svc.cluster.local` will currently match and return the result of an ExternalName service named `com` in the `default` namespace.

This change prevents this from happening.

### 2. Which issues (if any) are related?
fixes #6161

### 3. Which documentation changes (if any) need to be made?
none

### 4. Does this introduce a backward incompatible change or deprecation?
no, it fixes queries that should be failing but are returning an invalid response.

Commit:
plugin/kubernetes: filter ExternalName services from matching double subdomain wildcard

Signed-off-by: emaildanwilson <dan@controlplane.com>

remove double subdomain reference from review feedback

Signed-off-by: emaildanwilson <dan@controlplane.com>

not subdoman

Signed-off-by: emaildanwilson <dan@controlplane.com>

chrisohaver commented on plugin/kubernetes/handler_test.go:
```diff
@@ -408,6 +408,14 @@ var dnsTestCases = []kubeTestCase{
 			test.SOA("cluster.local.	5	IN	SOA	ns.dns.cluster.local. hostmaster.cluster.local. 1499347823 7200 1800 86400 5"),
 		},
 	}},
+	// A query for a double subdomain of an external service should not resolve to the external service
```
```suggestion
	// A query for subdomains of an external service should not resolve to the external service
```

emaildanwilson commented on plugin/kubernetes/handler_test.go:
```diff
@@ -408,6 +408,14 @@ var dnsTestCases = []kubeTestCase{
 	}},
+	// A query for a double subdomain of an external service should not resolve to the external service
+	{Case: test.Case{
+		Qname: "double.endpoint.external.testns.svc.cluster.local.", Qtype: dns.TypeCNAME,
```
I started with only setting the port/protocol pseudo segments if they started with an underscore but then went with this other approach because it could also effect headless service lookups. I can revisit this approach if it's preferred. hopefully some others can chime in on the SRV lookup questions.

chrisohaver commented on plugin/kubernetes/handler_test.go:
```diff
@@ -408,6 +408,14 @@ var dnsTestCases = []kubeTestCase{
 	}},
+	// A query for a double subdomain of an external service should not resolve to the external service
+	{Case: test.Case{
+		Qname: "double.endpoint.external.testns.svc.cluster.local.", Qtype: dns.TypeCNAME,
```
OK, the fix is OK as is.  Though I think the use of the terms `double` and `endpoint` is confusing here.  They are really just subdomains.  In fact, if the `endpoint` segment was actually being parsed as an "endpoint", the existing check would have caught it.

```suggestion
		Qname: "subdomain1.subdomain2.external.testns.svc.cluster.local.", Qtype: dns.TypeCNAME,
```

emaildanwilson replied:
sounds good. I'll make the updates to the names so it's less confusing.

chrisohaver commented on plugin/kubernetes/kubernetes.go:
```diff
@@ -446,7 +446,7 @@ func (k *Kubernetes) findServices(r recordRequest, zone string) (services []msg.
 		// External service
 		if svc.Type == api.ServiceTypeExternalName {
 			//External services cannot have endpoints, so skip this service if an endpoint is present in the request
```
```suggestion
			// External services do not have endpoints, nor can we accept port/protocol pseudo subdomains in an SRV query, so skip this service if endpoint, port, or protocol is non-empty in the request
```

chrisohaver commented:
FWIW, the title of the PR mentions "double wildcard", and the example query contains `*`'s, which used to haver a special nonstandard behavior in CoreDNS's _kubernetes_ plugin.

I thought I should clarify that CoreDNS no longer supports "wildcard" queries like this.  It treats a `*` segment as a literal `*`, per standard DNS behavior since 1.9.0.

It doesnt affect the validity of the issue or fix.  The issue exists for any number of subdomains > 1 .

chrisohaver commented on plugin/kubernetes/handler_test.go:
```diff
@@ -408,6 +408,14 @@ var dnsTestCases = []kubeTestCase{
 	}},
+	// A query for a subdomain of a subdomain of an external service should not resolve to the external service
+	{Case: test.Case{
+		Qname: "subdoman.subdomain.external.testns.svc.cluster.local.", Qtype: dns.TypeCNAME,
```
I know it doesn't matter,  but this will bother someone, might as well fix now.

```suggestion
		Qname: "subdomain.subdomain.external.testns.svc.cluster.local.", Qtype: dns.TypeCNAME,
```

emaildanwilson replied:
shoot. will fix.

emaildanwilson commented:
> FWIW, the title of the PR mentions "double wildcard", and the example query contains `*`'s, which used to have a special nonstandard behavior in CoreDNS's _kubernetes_ plugin.
> 
> I thought I should clarify that CoreDNS no longer supports "wildcard" queries like this. It treats a `*` segment as a literal `*`, per standard DNS behavior since 1.9.0.
> 
> It doesnt affect the validity of the issue or fix. The issue exists for any number of subdomains under the external service domain > 1 .

Thank you! I updated the title and added clarity to the description so that this doesn't get mixed up with wildcard dns query functionality.
